File: code/client/core/source/local-user-config/LocalUserConfig.cpp
﻿// <copyright file="LocalUserConfig.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <stdexcept>
#include <string>
#include <vector>

export module Soup.Core:LocalUserConfig;

import Opal;
import :RecipeValue;
import :SDKConfig;

using namespace Opal;

namespace Soup::Core
{
	/// <summary>
	/// The local user config container
	/// </summary>
	export class LocalUserConfig
	{
	private:
		static constexpr const char* Property_SDKs = "SDKs";

	public:
		/// <summary>
		/// Initializes a new instance of the <see cref="LocalUserConfig"/> class.
		/// </summary>
		LocalUserConfig() :
			_table()
		{
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="LocalUserConfig"/> class.
		/// </summary>
		LocalUserConfig(RecipeTable table) :
			_table(std::move(table))
		{
		}

		/// <summary>
		/// Gets or sets the list of SDKs
		/// </summary>
		bool HasSDKs()
		{
			return HasValue(Property_SDKs);
		}

		std::vector<SDKConfig> GetSDKs()
		{
			if (!HasSDKs())
				throw std::runtime_error("No SDKs.");

			auto& values = GetValue(Property_SDKs).AsList();
			auto result = std::vector<SDKConfig>();
			for (auto& value : values)
			{
				result.push_back(SDKConfig(value.AsTable()));
			}

			return result;
		}

		/// <summary>
		/// Raw access
		/// </summary>
		RecipeTable& GetTable()
		{
			return _table;
		}

		/// <summary>
		/// Equality operator
		/// </summary>
		bool operator ==(const LocalUserConfig& rhs) const
		{
			return _table == rhs._table;
		}

		/// <summary>
		/// Inequality operator
		/// </summary>
		bool operator !=(const LocalUserConfig& rhs) const
		{
			return !(*this == rhs);
		}

	private:
		bool HasValue(std::string_view key)
		{
			return _table.Contains(key.data());
		}

		const RecipeValue& GetValue(std::string_view key) const
		{
			const RecipeValue* value;
			if ( _table.TryGet(key.data(), value))
			{
				return *value;
			}
			else
			{
				throw std::runtime_error("Requested recipe value does not exist in the root table.");
			}
		}

	private:
		RecipeTable _table;
	};
}

File: code/generate-sharp/utilities/local-user-config/LocalUserConfig.cs
﻿// <copyright file="LocalUserConfig.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using System;
using System.Collections.Generic;

namespace Soup.Build.Utilities;

/// <summary>
/// The local user config container
/// </summary>
public class LocalUserConfig
{
	public static string Property_SDKs => "SDKs";

	/// <summary>
	/// Initializes a new instance of the <see cref="LocalUserConfig"/> class.
	/// </summary>
	public LocalUserConfig()
	{
		this.Document = new SMLDocument();
	}

	/// <summary>
	/// Initializes a new instance of the <see cref="LocalUserConfig"/> class.
	/// </summary>
	public LocalUserConfig(SMLDocument table)
	{
		this.Document = table;
	}

	/// <summary>
	/// Gets or sets the list of SDKs
	/// </summary>
	public bool HasSDKs()
	{
		return this.Document.Values.ContainsKey(Property_SDKs);
	}

	public IList<SDKConfig> GetSDKs()
	{
		if (this.Document.Values.TryGetValue(Property_SDKs, out var sdksValue))
		{
			var values = sdksValue.Value.AsArray();
			var result = new List<SDKConfig>();
			foreach (var value in values.Values)
			{
				result.Add(new SDKConfig(value.Value.AsTable()));
			}

			return result;
		}
		else
		{
			throw new InvalidOperationException("No SDKs.");
		}
	}

	public SDKConfig EnsureSDK(string name)
	{
		// Check the existing entries
		SMLArray? values;
		if (this.Document.Values.TryGetValue(Property_SDKs, out var sdksValue))
		{
			values = sdksValue.Value.AsArray();
			foreach (var value in values.Values)
			{
				var config = new SDKConfig(value.Value.AsTable());
				if (config.HasName() && config.Name == name)
					return config;
			}
		}
		else
		{
			values = this.Document.AddArrayWithSyntax(Property_SDKs);
		}

		// No matching SDK as a table array entry
		var sdkValueTable = values.AddTableWithSyntax(1);

		return new SDKConfig(sdkValueTable)
		{
			Name = name,
		};
	}

	/// <summary>
	/// Raw access
	/// </summary>
	public SMLDocument Document { get; }
}